src/main/java/com/google/cloud/dataproc/templates/dataplex/DataplexGCStoBQ.java

```diff
@@ -0,0 +1,391 @@
+   *
+   * @param partitionKeysList list of partition keys for Dataplex Entity
+   * @return a dataset with all distinct value of partition keys in BQ target table
+   */
+  private Dataset<Row> getBQTargetAvailablePartitionsDf(List<String> partitionKeysList) {
+    LOGGER.info("Reading target table: {}", targetTable);
+    spark.conf().set("viewsEnabled", "true");
+    spark.conf().set("materializationDataset", targetDataset);
+    try {
+      String sql =
+          String.format(
+              "select distinct %s FROM `%s`", String.join(",", partitionKeysList), targetTable);
+      return spark.read().format("bigquery").load(sql);
+    } catch (BigQueryConnectorException e) {
+      if (e.getCause().toString().contains("Not found: Table")) {
+        return null;
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Compare partitions in dataplex (partitionValuesInDataplex) with data available at BQ target
+   * table (bqTargetAvailablePartitionsDf) to identify new partitions
+   *
+   * @param partitionKeysList list of partition keys for Dataplex Entity
+   * @return a dataset with the GCS paths of new partitions
+   */
+  private Dataset<Row> getNewPartitionsPathsDS(
+      List<String> partitionKeysList,
+      Dataset<Row> dataplexPartitionsKeysDS,
+      Dataset<Row> bqPartitionsKeysDS) {
+    if (bqPartitionsKeysDS == null
+        || incrementalParittionCopy.equals(INCREMENTAL_PARTITION_COPY_NO)) {
+      return dataplexPartitionsKeysDS.select("__gcs_location_path__");
+    }
+
+    dataplexPartitionsKeysDS.createOrReplaceTempView("dataplexPartitionsKeysDS");
+    bqPartitionsKeysDS.createOrReplaceTempView("bqPartitionsKeysDS");
+    String joinClause =
+        partitionKeysList.stream()
+            .map(str -> String.format("t1.%s=t2.%s", str, str))
+            .collect(Collectors.joining(" AND "));
+    Dataset<Row> newPartitionsDS =
+        spark.sql(
+            "SELECT __gcs_location_path__ "
+                + "FROM dataplexPartitionsKeysDS t1 "
+                + "LEFT JOIN bqPartitionsKeysDS t2 ON "
+                + joinClause
+                + " WHERE t2.id is null");
+    return newPartitionsDS;
+  }
+
+  /**
+   * Loads from GCS all new partitions
+   *
+   * @param newPartitionsPathsDf a dataset with the GCS paths of new partitions
+   * @return a dataset with the GCS paths of new partitions
+   */
+  private Dataset<Row> getNewPartitionsDS(Dataset<Row> newPartitionsPathsDf) {
+    Row[] result = (Row[]) newPartitionsPathsDf.select("__gcs_location_path__").collect();
+    Dataset<Row> newPartitionsDS = null;
+    for (Row row : result) {
+      String path = row.get(0).toString() + "/*";
+      LOGGER.info("Loading data from GCS path: {}", path);
+      Dataset<Row> newPartitionTempDS =
+          sqlContext
+              .read()
+              .format(inputFileFormat)
+              .option(GCS_BQ_CSV_HEADER, true)
+              .option(GCS_BQ_CSV_INFOR_SCHEMA, true)
```

Comment:
is this true for all file formats?

Reply:
These are only necessary for CSV, but I tested for Parquet, Avro and JSON and it seems like spark will just ignore these options when it is not reading CSV or rather these options have no effect.

I don't mind having unnecessary options when reading formats that are not CSV, since it causes no harm and actually makes the code more readable.

Reply:
sg